File: src/main/java/com/vaadin/demo/component/badge/BadgeHighlight.java
package com.vaadin.demo.component.badge;

import com.vaadin.demo.DemoExporter; // hidden-source-line
import com.vaadin.demo.domain.DataService;
import com.vaadin.demo.domain.Report;
import com.vaadin.flow.component.grid.Grid;
import com.vaadin.flow.component.html.Div;
import com.vaadin.flow.component.html.Span;
import com.vaadin.flow.data.renderer.LocalDateRenderer;
import com.vaadin.flow.router.Route;

import java.time.LocalDate;
import java.time.ZoneId;
import java.util.List;

@Route("badge-highlight")
public class BadgeHighlight extends Div {

    public BadgeHighlight() {
        Grid<Report> grid = new Grid<>(Report.class, false);

        grid.addColumn(Report::getReport).setHeader("Report");
        grid.addColumn(
                new LocalDateRenderer<>(BadgeHighlight::getDueDate, "MMM d, YYYY"))
                .setHeader("Due date");
        grid.addColumn(Report::getAssignee).setHeader("Assignee");
        // tag::snippet1[]
        grid.addComponentColumn(report ->
                createStatusBadge(report.getStatus()))
                .setHeader("Status");
        // end::snippet1[]

        List<Report> reports = DataService.getReports();
        grid.setItems(reports);

        add(grid);
    }

    // tag::snippet2[]
    private Span createStatusBadge(String status) {
        String theme;
        switch (status) {
            case "In progress":
                theme = "badge primary";
                break;
            case "Completed":
                theme = "badge success primary";
                break;
            case "Cancelled":
                theme = "badge error primary";
                break;
            default:
                theme = "badge contrast primary";
                break;
        }
        Span badge = new Span(status);
        badge.getElement().getThemeList().add(theme);
        return badge;
    }
    // end::snippet2[]

    private static LocalDate getDueDate(Report report) {
        return report.getDue().toInstant().atZone(ZoneId.systemDefault())
                .toLocalDate();
    }
    public static class Exporter extends DemoExporter<BadgeHighlight> { // hidden-source-line
    } // hidden-source-line
}

File: frontend/demo/component/badge/badge-highlight.ts
import { getReports, ReportStatus } from 'Frontend/demo/domain/DataService'; // hidden-source-line
import 'Frontend/demo/init'; // hidden-source-line
import type Report from 'Frontend/generated/com/vaadin/demo/domain/Report'; // hidden-source-line
import { html, LitElement } from 'lit';
import { customElement, state } from 'lit/decorators.js';
import '@vaadin/grid';
import { columnBodyRenderer } from '@vaadin/grid/lit.js';
import '@vaadin/icon';
import '@vaadin/icons';
import { applyTheme } from 'Frontend/generated/theme';

const dateFormatter = new Intl.DateTimeFormat('en-US', {
  year: 'numeric',
  month: 'short',
  day: 'numeric',
});

@customElement('badge-highlight')
export class Example extends LitElement {
  @state()
  private items: readonly Report[] = [];

  protected override createRenderRoot() {
    const root = super.createRenderRoot();
    // Apply custom theme (only supported if your app uses one)
    applyTheme(root);
    return root;
  }

  protected override async firstUpdated() {
    this.items = await getReports();
  }

  protected override render() {
    // tag::snippet[]
    return html`
      <vaadin-grid .items="${this.items}">
        <vaadin-grid-column path="report" header="Report"></vaadin-grid-column>
        <vaadin-grid-column
          header="Due date"
          ${columnBodyRenderer<Report>(
            ({ due }) => html`${dateFormatter.format(new Date(due))}`,
            []
          )}
        ></vaadin-grid-column>
        <vaadin-grid-column path="assignee" header="Assignee"></vaadin-grid-column>
        <vaadin-grid-column
          header="Status"
          ${columnBodyRenderer<Report>(({ status }) => {
            let title: string;
            let theme: string;

            switch (status) {
              case ReportStatus.COMPLETED:
                title = 'Completed';
                theme = 'success';
                break;
              case ReportStatus.IN_PROGRESS:
                title = 'In progress';
                theme = '';
                break;
              case ReportStatus.CANCELLED:
                title = 'Cancelled';
                theme = 'error';
                break;
              default:
                title = 'On hold';
                theme = 'contrast';
                break;
            }

            return html`<span theme="badge ${theme} primary">${title}</span>`;
          }, [])}
        ></vaadin-grid-column>
      </vaadin-grid>
    `;
    // end::snippet[]
  }
}
